Touch Bar Simulator/TouchBarWindow.swift

```diff
@@ -45,6 +66,177 @@ final class TouchBarWindow: NSPanel {
+				self.setIsVisible(true)
+			}
+		}
+	}
+
+	@objc
+	func handleDockBehavior() {
+		guard
+			self.docking != nil,
+			let visibleFrame = NSScreen.main?.visibleFrame,
+			let screenFrame = NSScreen.main?.frame
+		else {
+			return
+		}
+
+		var detectionRect: NSRect = .zero
+		if self.docking! == .dockedToBottom {
+			if self.isVisible {
+				detectionRect = CGRect(
+					x: 0,
+					y: 0,
+					width: visibleFrame.width,
+					height: self.frame.height + (screenFrame.height - visibleFrame.height - NSStatusBar.system.thickness)
+				)
+			} else {
+				detectionRect = CGRect(x: 0, y: 0, width: visibleFrame.width, height: 1)
+			}
+		} else if self.docking! == .dockedToTop {
+			if self.isVisible {
+				detectionRect = CGRect(
+					x: 0,
+					// Without `+ 1`, the Touch Bar would glitch (toggling rapidly).
+					y: screenFrame.height - self.frame.height - NSStatusBar.system.thickness + 1,
+					width: visibleFrame.width,
+					height: self.frame.height + NSStatusBar.system.thickness)
+			} else {
+				detectionRect = CGRect(
+					x: 0,
+					y: screenFrame.height,
+					width: visibleFrame.width,
+					height: 1)
+			}
+		}
+
+		let mouseLocation = NSEvent.mouseLocation
+		if detectionRect.contains(mouseLocation) {
+			dismissAnimationDidRun = false
+
+			guard
+				!showTouchBarTimer.isValid,
+				!showAnimationDidRun
+			else {
+				return
+			}
+
+			showTouchBarTimer = Timer.scheduledTimer(withTimeInterval: 0.3, repeats: false) { _ in
+				self.showTouchBarWithAnimation()
+				self.showAnimationDidRun = true
+			}
+		} else {
+			showTouchBarTimer.invalidate()
+			showTouchBarTimer = Timer()
+			showAnimationDidRun = false
+
+			if self.isVisible && !dismissAnimationDidRun {
+				dismissTouchBarWithAnimation()
+				dismissAnimationDidRun = true
+			}
+		}
+	}
+
+	func moveToStartPoint() {
+		guard self.docking != nil else {
+			return
+		}
+
+		if self.docking! == .dockedToTop {
+			self.moveTo(x: .center, y: .top)
+		} else if self.docking! == .dockedToBottom {
+			self.moveTo(x: .center, y: .bottom)
+		}
+	}
+
+	var showAnimationDidRun = false
+	var dismissAnimationDidRun = false
+
+	func showTouchBarWithAnimation() {
+		guard
+			self.docking != nil &&
+			self.docking! == .dockedToTop ||
+			self.docking! == .dockedToBottom
+		else {
+			return
+		}
+
+		var startOrigin: CGPoint!
+		let endFrame = self.frame
+		self.setIsVisible(true)
+		if self.docking! == .dockedToTop {
+			startOrigin = CGPoint(x: self.frame.origin.x, y: self.frame.origin.y + self.frame.height)
+		} else if self.docking! == .dockedToBottom {
+			startOrigin = CGPoint(x: self.frame.origin.x, y: 0 - self.frame.height)
+		}
+
+		self.setFrameOrigin(startOrigin)
+
+		NSAnimationContext.runAnimationGroup({ context in
+			context.duration = TimeInterval(0.3)
+			self.animator().setFrame(endFrame, display: false, animate: true)
+		}, completionHandler: {
+			self.moveToStartPoint()
+		})
+	}
+
+	func dismissTouchBarWithAnimation() {
+		guard
+			self.docking != nil &&
```

Comment:
This should be properly unwrapped. You should generally never use `!`. This applies to all the places you use `docking!`.